docs/source/more_types.rst

```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+When you call an overloaded function, mypy will infer the correct return
+type by picking the best matching variant, after taking into consideration
+both the argument types and arity. However, a call is never type
+checked against the implementation. This is why mypy will report calls
+like ``ip_address(1, 2)`` as being invalid even though it matches the
+implementation signature.
+
+If there are multiple equally good matching variants, mypy will select
+the variant that was defined first. For example, consider the following
+program:
+
+.. code-block:: python
+
+    from typing import List, overload
+
+    @overload
+    def summarize(data: List[str]) -> str: ...
+
+    @overload
+    def summarize(data: List[int]) -> int: ...
+
+    def summarize(data):
+        # ...snip...
+
+    # What is the type of 'output'? str or int?
+    output = summarize([])
+
+The ``summarize([])`` call matches both variants: an empty list could
+be either a ``List[str]`` or a ``List[int]``. In this case, mypy
+will break the tie by picking the first matching variant: ``output``
+will have an inferred type of ``str``. The implementor is responsible
+for making sure ``summarize`` breaks ties in the same way at runtime.
+
+There are however are two exceptions to the "pick the first match" rule.
+First, if multiple variants match due to an argument being of type
+``Any``, mypy will make the inferred type also be ``Any``:
+
+.. code-block:: python
+
+    dynamic_var: Any = some_dynamic_function()
+
+    # output2 is of type 'Any'
+    output2 = summarize(dynamic_var)
+
+Second, if multiple variants match due to one or more of the arguments
+being a union, mypy will make the inferred type be the union of the 
+matching variant returns:
+
+.. code-block:: python
+
+    some_list: Union[List[str], List[int]]
+
+    # output3 is of type 'Union[str, int]'
+    output3 = summarize(some_list)
+
+.. note::
+
+   Due to the "pick the first match" rule, changing the order of your
+   overload variants can change how mypy type checks your program.
+
+   To minimize potential issues, we recommend ordering your variants
+   from most to least specific. Your implementation should also
+   perform runtime checks (e.g. ``isinstance`` checks) in the same
+   order as the listed variants.
+
+   If your variants have no inherent relationship to each other, they
+   can naturally be listed in any arbitrary order.
+
+Type checking the variants
+--------------------------
+
+Although the "pick the first match" algorithm works well in many cases, it can
+sometimes make mypy infer the incorrect type. For example, consider the following
+overload definition:
+
+.. code-block:: python
+
+    from typing import overload, Union
+
+    @overload
+    def unsafe_func(x: int) -> int: ...
+
+    @overload
+    def unsafe_func(x: object) -> str: ...
+
+    def unsafe_func(x: object) -> Union[int, str]:
+        if isinstance(x, int):
+            return x
+        else:
+            return "Unsafe!"
```

Comment:
It is not clear for a typical reader why it is unsafe at this point, use a different word here, maybe "Hm, not really a number".